lib/active_merchant/billing/gateways/trexle.rb

```diff
@@ -0,0 +1,217 @@
+
+        add_creditcard(post, creditcard)
+        add_customer_data(post, options)
+        add_address(post, creditcard, options)
+        commit(:put, "customers/#{CGI.escape(token)}", post, options)
+      end
+
+      def supports_scrubbing
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(/(number\\?":\\?")(\d*)/, '\1[FILTERED]').
+          gsub(/(cvc\\?":\\?")(\d*)/, '\1[FILTERED]')
+      end
+      private
+
+      def add_amount(post, money, options)
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+        post[:currency] = post[:currency].upcase if post[:currency]
+      end
+
+      def add_customer_data(post, options)
+        post[:email] = options[:email] if options[:email]
+        post[:ip_address] = options[:ip] if options[:ip]
+      end
+
+      def add_address(post, creditcard, options)
+        return if creditcard.kind_of?(String)
+        address = (options[:billing_address] || options[:address])
+        return unless address
+
+        post[:card] ||= {}
+        post[:card].merge!(
+          address_line1: address[:address1],
+          address_line2: address[:address_line2],
+          address_city: address[:city],
+          address_postcode: address[:zip],
+          address_state: address[:state],
+          address_country: address[:country]
+        )
+      end
+
+      def add_invoice(post, options)
+        post[:description] = options[:description] || "Active Merchant Purchase"
+      end
+
+      def add_creditcard(post, creditcard)
+        if creditcard.respond_to?(:number)
+          post[:card] ||= {}
+
+          post[:card].merge!(
+            number: creditcard.number,
+            expiry_month: creditcard.month,
+            expiry_year: creditcard.year,
+            cvc: creditcard.verification_value,
+            name: creditcard.name
+          )
+        elsif creditcard.kind_of?(String)
+          if creditcard =~ /^token_/
+            post[:card_token] = creditcard
+          else
+            post[:customer_token] = creditcard
+          end
+        end
+      end
+
+      def headers(params = {})
+        result = {
+          "Content-Type" => "application/json",
+          "Authorization" => "Basic #{Base64.strict_encode64(options[:api_key] + ':').strip}"
+        }
+
+        result['X-Partner-Key'] = params[:partner_key] if params[:partner_key]
+        result['X-Safe-Card'] = params[:safe_card] if params[:safe_card]
+        result
+      end
+      
+      def commit(method, action, params, options)
+        url = "#{test? ? test_url : live_url}/#{action}"
+        raw_response = ssl_request(method, url, post_data(params), headers(options))
+        parsed_response = parse(raw_response)
+        success_response(parsed_response) 
+      rescue ResponseError => e
+        error_response(parse(e.response.body))
```

Comment:
@hossamhossny apologies on the tardy replies- lots going on these days. 

WePay and Pin have been in Active Merchant a while so they're not following more recent patterns. But, given that we're not as explicit about what we're looking for or conventions I'll let this one in as is. @hossamhossny apologies on the tardy replies- lots going on these days. 

WePay and Pin have been in Active Merchant a while so they're not following more recent patterns. But, given that we're not as explicit about what we're looking for or conventions I'll let this one in as is. 

If you wanna squash your commits and rebase on `master`@hossamhossny apologies on the tardy replies- lots going on these days. 

WePay and Pin have been in Active Merchant a while so they're not following more recent patterns. But, given that we're not as explicit about what we're looking for or conventions I'll let this one in as is. 

If you wanna squash all your commits and rebase on master I can merge in.